pkg/asset/machines/clusterapi.go

```diff
@@ -227,24 +227,18 @@ func (c *ClusterAPI) Generate(ctx context.Context, dependencies asset.Parents) e
 		}
 		client := icazure.NewClient(session)
 
-		if len(mpool.Zones) == 0 {
-			// if no azs are given we set to []string{""} for convenience over later operations.
-			// It means no-zoned for the machine API
-			mpool.Zones = []string{""}
-		}
 		if len(mpool.Zones) == 0 {
 			azs, err := client.GetAvailabilityZones(ctx, ic.Platform.Azure.Region, mpool.InstanceType)
 			if err != nil {
 				return fmt.Errorf("failed to fetch availability zones: %w", err)
 			}
 			mpool.Zones = azs
-			if len(azs) == 0 {
-				// if no azs are given we set to []string{""} for convenience over later operations.
-				// It means no-zoned for the machine API
-				mpool.Zones = []string{""}
-			}
 		}
-		// client.GetControlPlaneSubnet(ctx, ic.Platform.Azure.ResourceGroupName, ic.Platform.Azure.VirtualNetwork, )
+		if len(mpool.Zones) == 0 {
+			// if no azs are given we set to []string{""} for convenience over later operations.
+			// It means no-zoned for the machine API
+			mpool.Zones = []string{""}
+		}
```

Comment:
I'm confused by this change. Why do the earlier conditional if we're just going to overwrite it here?

/hold

Reply:
@patrickdillon some regions do not support availability zones. So the call to `client.GetAvailabilityZones` can still return an empty slice/array.

Reply:
There is not really any overriding unless `len(mpool.Zones) == 0`. That being said, this was a workaround needed by MAPI for the non-zonal case. Maybe we don't need this hack anymore for CAPZ?

Reply:
Oh, it could still be necessary to set this anyway for CPMS/MAPI.

Reply:
I don't understand how this fixes the problem, but masters do belong to separate azs in my testing. So I don't want to block too much, but the code also looks off to me unless I'm missing something:

> So the call to client.GetAvailabilityZones can still return an empty slice/array.

The assignment of that value `mpool.Zones = azs` will always be overwritten in the code change. There are two matching sequential conditionals so it would make sense to combine them right?

https://github.com/openshift/installer/blob/4a0fb6b2821b86eead45815b02cfc055b9149568/pkg/asset/machines/clusterapi.go#L230-L241

Reply:
@patrickdillon The logic is:
1. Use the Zones set in the machine pool (zones specified in the install-config.yaml)
2. If no zones set, then try to discover zones for that instance type in the configured region
3. If still no zones are set, then add `""` to the zones list to mean "no-zonal" for MAPI.

We can move one `if` inside the other if that makes it easier to read:
```go
if len(mpool.Zones) == 0 { 
 	azs, err := client.GetAvailabilityZones(ctx, ic.Platform.Azure.Region, mpool.InstanceType) 
 	if err != nil { 
 		return fmt.Errorf("failed to fetch availability zones: %w", err) 
 	}
    if len(azs) > 0 {
 	    mpool.Zones = azs
    } else {
        // if no azs are given we set to []string{""} for convenience over later operations. 
 	    // It means no-zoned for the machine API 
 	    mpool.Zones = []string{""}
    } 
}
```
but the effect is the same.

Reply:
> I don't understand how this fixes the problem

What actually fixes the problem is the removal of these lines https://github.com/openshift/installer/pull/8817/files#diff-c8a93a8c9fb0e5dbfa50e3a8aa1fbd253dd324fb981a01a69f6bf843305117cbL230-L234. See this comment https://github.com/openshift/installer/pull/8817#discussion_r1711664884 for an explanation of why.

Reply:
ack makes sense now and sorry i missed the earlier thread!

/hold cancel